Button in Metabase Admin section to purchase Metabot AI add-on

devurandom opened the pull request:
== Goal ==

In case the Metabase instance does not have the Metabot AI add-on, we
want to show a page in the admin section to purchase it.

== How to test ==

1. Pick a valid token from staging with the `offer-metabase-ai` feature, but without the `metabot-v3` feature (`metabase-ai` add-on)
2. Pick the token's instance's Metabase API key
3. Pick an email address that matches the store user of the token's organization
4. Spin up a Metabase instance with that API key, token and user
5. Navigate to the admin section, AI tab
6. See the new page to purchase the add-on and click the button

Closes: https://linear.app/metabase/issue/CLO-4370/button-in-metabase-admin-section-to-purchase-metabot-ai-add-on

devurandom committed:
Button in Metabase Admin section to purchase Metabot AI add-on

== Goal ==

In case the Metabase instance does not have the Metabot AI add-on, we
want to show a page in the admin section to purchase it.

== How to test ==

1. Spin up a Metabase instance with a valid token
2. Navigate to the admin section, AI tab
3. See the new page to purchase the add-on and click the button

Closes: https://linear.app/metabase/issue/CLO-4370/button-in-metabase-admin-section-to-purchase-metabot-ai-add-on

andsalves commented on enterprise/frontend/src/metabase-enterprise/metabot/components/MetabotAdmin/MetabotAdminPurchase.tsx:
```diff
@@ -0,0 +1,234 @@
+                    fw="bold"
+                  >{t`You won't be charged automatically.`}</Text>{" "}
+                  {t`After the trial, you can decide if you'd like to subscribe.`}
+                </Text>
```
Missing space after "Your first month is on us."

devurandom replied:
https://github.com/metabase/metabase/pull/62513/commits/5072708b8e2c221730bb9f40f3105cbdd41abad9

andsalves commented on enterprise/frontend/src/metabase-enterprise/metabot/components/MetabotAdmin/MetabotAdminPurchase.tsx:
```diff
@@ -0,0 +1,234 @@
+        typeof error.errors.terms_of_service === "string")));
+
+export const handleFieldError = (error: unknown) => {
+  if (isFieldError(error)) {
+    if (typeof error === "string") {
+      throw { data: { errors: { terms_of_service: error } } };
+    } else if ("message" in error) {
+      throw { data: { errors: { terms_of_service: error.message } } };
+    } else if ("errors" in error) {
+      throw { data: error };
+    }
+  }
```
Nitpick:
We could avoid nested if`s
```suggestion
  if (!isFieldError(error)) {
    return;
  }

  if (typeof error === "string") {
    throw { data: { errors: { terms_of_service: error } } };
  }
  
  if ("message" in error) {
    throw { data: { errors: { terms_of_service: error.message } } };
  }
  
  if ("errors" in error) {
    throw { data: error };
  }
```

andsalves commented on enterprise/frontend/src/metabase-enterprise/metabot/components/MetabotAdmin/MetabotAdminPurchase.tsx:
```diff
@@ -0,0 +1,234 @@
+  terms_of_service: boolean;
+}
+
+const MetabotAdminSettingUpModal = ({
```
Can we please move this to its own file

andsalves commented on enterprise/frontend/src/metabase-enterprise/metabot/components/MetabotAdmin/MetabotAdminPurchase.tsx:
```diff
@@ -0,0 +1,234 @@
+        terms_of_service,
+      })
+        .unwrap()
+        .then()
```
calling `.then()` is unnecessary

andsalves reviewed (changes requested):
A few comments, nothing much. 
Still reviewing

devurandom committed:
Only offer the add-on to Cloud customers

devurandom committed:
Add frontend tests

devurandom committed:
Split MetabaseAdminPurchase.tsx into two files

devurandom committed:
Flatten conditions in handleFieldError

devurandom committed:
Add missing whitespace

devurandom committed:
Ensure errors raised by handleFieldError are visible

devurandom committed:
Shorten local variable name

devurandom committed:
Respond with HTTP status 500 when Store API request fails

`{:status 500}` is rendered as a response with status 200 and a body of
`{"status": 500}`, wheras `{:status 500 :body ""}` is rendered as a
response with status 500 with empty body.

devurandom committed:
Handle HTTP status 403 and 401 from Store API

devurandom committed:
Show "Success" on FormSubmitButton only once the HTTP response is received

devurandom committed:
Limit offer to select subscriptions only

devurandom committed:
Offer add-on only to Metabase Store users

devurandom commented on enterprise/backend/src/metabase_enterprise/cloud_add_ons/api.clj:
```diff
@@ -0,0 +1,37 @@
+        (case (-> e ex-data :status)
+          404 {:status 404 :body "Could not establish a connection to Metabase Cloud."}
+          400 {:status 400 :body "Could not purchase this add-on."}
+          {:status 500})))))
```
https://github.com/metabase/metabase/pull/62513/commits/67a31360256bd6b70ab9f2f8c8c7f8bca04cf4fd

chodorowicz reviewed (changes requested):
Cool! Looks pretty good already. I've couple of comments and we need to handle token invalidation.

devurandom committed:
Refresh token features after purchase

devurandom committed:
Fix admin paths and navigation when add-on is present

devurandom committed:
Fix missing error indicator when respoding with 500 Internal Server Error

devurandom committed:
Translate user-visible error strings

devurandom committed:
Use Mantine's useDisclosure instead of React's useState

devurandom committed:
Move Metabot to 3rd position in admin navigation bar

devurandom committed:
Translate image alt text

devurandom committed:
Import SVG icon instead of referencing it directly

devurandom commented on enterprise/frontend/src/metabase-enterprise/metabot/components/MetabotAdmin/MetabotPurchaseSettingUpModal.tsx:
```diff
@@ -0,0 +1,74 @@
+  >
+    <Stack align="center" gap="lg" my="4.5rem">
+      <Box h={96} pos="relative" w={96}>
+        <img src="app/assets/img/metabot-cloud-96x96.svg" alt="Metabot Cloud" />
```
https://github.com/metabase/metabase/pull/62513/commits/2ff3eca17e8cc1e79acc1597948f8fd60a924e19

devurandom committed:
Use Mantine Image component instead of raw HMTL img

devurandom committed:
Token feature will be added in v56

chodorowicz commented on enterprise/frontend/src/metabase-enterprise/metabot/components/MetabotAdmin/MetabotPurchasePage.tsx:
```diff
@@ -0,0 +1,181 @@
+              <Text component="span" fw="bold">{t`Explain queries`}</Text>{" "}
+              <Text component="span">{t`or existing charts`}</Text>
+            </List.Item>
+          </List>
```
No idea 😅 I haven't seen one.

devurandom committed:
Embed the Metabase AI demo video from metabase.com

Co-authored-by: Roman Abdulmanov <roman@metabase.com>

devurandom committed:
Permit use of metabase.util from metabase-enterprise.cloud-add-ons

devurandom committed:
Make empty state alternate text bold for visibility and s/Store user/Store Admin/

devurandom committed:
Move eslint comment right next to the text it affects to prevent auto-formatting from moving the two apart and causing linter failures

devurandom committed:
Offer the add-on also on v55

chodorowicz commented on enterprise/frontend/src/metabase-enterprise/metabot/components/MetabotAdmin/MetabotPurchasePage.tsx:
```diff
@@ -0,0 +1,200 @@
+  const [purchaseMetabotAddOn] = usePurchaseMetabotCloudAddOnMutation();
+  const onSubmit = useCallback(
+    async ({ terms_of_service }: MetabotPurchaseFormFields) => {
+      await purchaseMetabotAddOn({
```
When I was testing this I had to wait a bit before modal opened, and when it opened then Metabot was immediately enabled. Aren't we waiting [for both enabling the Metabot AI and refreshing the token](https://github.com/metabase/metabase/pull/62513/files#diff-ca407010a35905a2fcfbccf6a0900b8f89d9a045b12ee7216c4eab08a9383e98R38). Maybe we should open the modal right away and present the loading state in modal, instead of showing the blocked button for extended time.

devurandom replied:
Should we then do something about the "Success" text on the `FormSubmitButton`?  It appears a bit odd to me to have the form greyed out and covered by a modal that says "wait" and in the background see a green "Success" light up.

devurandom replied:
> Aren't we waiting for both enabling the Metabot AI and refreshing the token.

We wait only for purchasing the add-on and dropping the token cache.  Then the frontend triggers another token check when showing the modal: https://github.com/metabase/metabase/pull/62513/files#diff-c657ec9cd4c5e1b8fa7ec286927a67fe2421a010bcfae333c036c705b5db3e12R23

devurandom replied:
Showing the modal earlier: https://github.com/metabase/metabase/pull/62513/commits/31ddb144831b0fd9b037504388c8a8a68671de1d

chodorowicz replied:
Nice! Better UX ☺️

devurandom committed:
Show a Store Admin email address, if known

devurandom commented on enterprise/frontend/src/metabase-enterprise/metabot/components/MetabotAdmin/MetabotPurchasePage.tsx:
```diff
@@ -0,0 +1,181 @@
+          </FormProvider>
+        ) : (
+          /* eslint-disable-next-line no-literal-metabase-strings -- This string only shows for admins. */
+          <Text>{t`Please ask a Metabase Store user of your organization to enable this for you.`}</Text>
```
@trinya approved this version of the empty state alternate text (at the bottom, below the divider): 
<img width="1796" height="1802" alt="image" src="https://github.com/user-attachments/assets/ad2537e3-f837-42a4-9ba3-4dfa79cb2b48" />
https://github.com/metabase/metabase/pull/62513/commits/ca77ef3ca8e67bbde25c0541edfd7e4ec6ea0dd1

devurandom committed:
Use default Metabase box shadow

devurandom commented on enterprise/frontend/src/metabase-enterprise/metabot/components/MetabotAdmin/MetabotPurchaseSettingUpModal.tsx:
```diff
@@ -0,0 +1,74 @@
+          style={{
+            borderRadius: "100%",
+            // eslint-disable-next-line no-color-literals
+            boxShadow: `0 0 0 1px rgba(0, 0, 0, 0.05), 0 1px 6px 0 rgba(0, 0, 0, 0.10)`,
```
Thanks a lot! https://github.com/metabase/metabase/pull/62513/commits/d801e8b5b6abd111a46d5e27889fcdd063c16185

devurandom committed:
Show modal earlier, before making the HTTP request to the backend

devurandom committed:
Exclude offer-metabase-ai token feature from stats ping

chodorowicz reviewed (approved):
Great jobs! This works as expected. But let's wait for sure for more BE focused review.

devurandom committed:
Adjust frontend test to string change

dpsutton reviewed (changes requested):
the backend looks right but there are 0 tests on this new feature. We should fix that. Also, looking at the harbor master client, it still reaches out and uses `(cloud-migration/store-api-url)`. perhaps the cloud migration should use some harbor master endpoint, instead of the client using a cloud migration endpoint.

andsalves reviewed (approved):
It looks good to me!

devurandom committed:
Move store-api-url setting into its own module

This way `metabase-enterprise.harbormaster.client` and
`metabase.cloud-migration.models.cloud-migration` can depend on it.

devurandom committed:
Move error message vars from local to file scope

This works, because `deferred-tru` returns an object that is translated
when `toString` is invoked on it, not when it is created.

devurandom committed:
Add backend tests

devurandom committed:
Also test clicking the button in frontend tests

devurandom committed:
Refresh token features more quickly

devurandom committed:
Make error string vars private

devurandom committed:
Declare store-api module

devurandom committed:
Test frontend making a HTTP request

devurandom committed:
Resolve "metabase.store-api.init should require metabase.store-api.settings"

... by adopting the same file structure as `metabase.cloud-migration.*`

devurandom committed:
Fix module declaration

Fixes: 18f62cdf37d59a884cb3458a9956ee4270e03d63

devurandom committed:
Fix "metabase.core.init should require metabase.store-api.init"

Fixes: 18f62cdf37d59a884cb3458a9956ee4270e03d63

devurandom committed:
Fix "Namespace metabase.store-api.init is not an allowed external API namespace for the store-api module"

devurandom commented:
Thanks for the reviews!

> the backend looks right but there are 0 tests on this new feature. We should fix that.

I added tests for the pre-conditions and error responses in https://github.com/metabase/metabase/pull/62513/commits/fc34046e699465286ad9ad77e4bacb976d3f451f and for the HTTP request made by the frontend in https://github.com/metabase/metabase/pull/62513/commits/81547636e2fc4ca9de60df351399f126502e72de.

> Also, looking at the harbor master client, it still reaches out and uses `(cloud-migration/store-api-url)`. perhaps the cloud migration should use some harbor master endpoint, instead of the client using a cloud migration endpoint.

Since `metabase-enterprise.harbormaster.client` is an enterprise module while `metabase.cloud-migration.*` is a non-enterprise module, the latter cannot depend on the former.  I moved the setting into its own module in https://github.com/metabase/metabase/pull/62513/commits/6c8efe002a477e03a3c53992473327ced24f246b to resolve this (and then fixed a linter error by changing the file structure in https://github.com/metabase/metabase/pull/62513/commits/18f62cdf37d59a884cb3458a9956ee4270e03d63).

iethree reviewed (changes requested):
we should address the localization issues before merging

devurandom committed:
Use "await ... find" instead of "waitFor ... get"

devurandom committed:
Combine translated strings for more context during translation

devurandom commented on enterprise/frontend/src/metabase-enterprise/metabot/components/MetabotAdmin/MetabotPurchasePage.unit.spec.tsx:
```diff
@@ -0,0 +1,137 @@
+      expect(
+        screen.getByText(/Setting up Metabot AI, please wait/),
+      ).toBeInTheDocument();
+    });
```
https://github.com/metabase/metabase/pull/62513/commits/45eb0435a77bc4d88e91c692b24982180d397d0f

devurandom commented on enterprise/frontend/src/metabase-enterprise/metabot/components/MetabotAdmin/MetabotPurchasePage.tsx:
```diff
@@ -0,0 +1,201 @@
+              <Text component="span">{
+                /* eslint-disable-next-line no-literal-metabase-strings -- This string only shows for admins. */
+                t`in the Metabase SQL editor`
+              }</Text>
```
https://github.com/metabase/metabase/pull/62513/commits/2824018380b0ef9536774ce8d089840d1b84970c

devurandom committed:
Fix "Warning: An update to MetabotPurchaseSettingUpModal inside a test was not wrapped in act(...)."

devurandom committed:
Remove unnecessary router from test

iethree reviewed (approved):
❤️

devurandom committed:
Don't poll token features when modal is not yet opened

devurandom committed:
Close modal on error

Fixes: 31ddb144831b0fd9b037504388c8a8a68671de1d

dpsutton reviewed (approved):
approving with notes:
- we should probably mark who requested this feature. It would be nice if there is ever any confusion we can say who and when changed the billing
- we should log errors when the upgrade doesn't work. right now we return specific 404,403,401,400 errors. We probably want a bit more info there.
- the tests are really verbose. I think you could make them more concise, maybe a doseq over some map of the status codes to responses like`error-no-connection`. I suspect the tests could be much smaller while giving the same expressivity.

devurandom committed:
Pass information about Metabase user to Store API

Suggested-by: Dan Sutton <dan@metabase.com>
References: https://metaboat.slack.com/archives/C08N6CX5EMQ/p1756143160721289?thread_ts=1755790723.074589&cid=C08N6CX5EMQ

devurandom committed:
Mark post-product-type-test sequential

devurandom committed:
Log exception from Store API

devurandom committed:
Make error pass through tests more concise

devurandom committed:
Empty bump

devurandom committed:
Pass flag to useTokenRefreshUntil instead of unmounting the modal

Modals in Metabase have to always stay in the DOM, so we pass a boolean
flag through  `useTokenRefreshUntil` into `useEffect` to prevent
reloading the session-properties from the backend.

Suggested-by: Anderson Freitas <anderson@metabase.com>

devurandom committed:
Fix test failure

Fixes: e0562a4a40e60b84e0904f72b47eb772f2823ddd
